File: docs/components_page/components/popover/direction.jl
using DashBootstrapComponents


function make_popover(placement)
    return dbc_popover(
        [dbc_popoverheader("Header"), dbc_popoverbody("This is a $placement popover")],
        id = "popover-$placement",
        target = "popover-$placement-target",
        placement = placement,
        is_open = false,
    )
end;

function make_button(placement)
    return dbc_button(
        "Popover on $placement",
        id = "popover-$placement-target",
        className = "mx-2",
        n_clicks = 0,
    )
end;


popovers = html_div(
    [
        [make_button(p) for p in ["top", "left", "right", "bottom"]]
        [make_popover(p) for p in ["top", "left", "right", "bottom"]]
    ],
);

for p in ["top", "left", "right", "bottom"]
    callback!(
        app,
        Output("popover-$p", "is_open"),
        Input("popover-$p-target", "n_clicks"),
        State("popover-$p", "is_open"),
    ) do n, is_open
        return n > 0 ? is_open == 0 : is_open
    end
end;

File: docs/components_page/components/popover/direction.R
library(dashBootstrapComponents)

make_popover <- function(placement) {
  return(
    dbcPopover(
      list(
        dbcPopoverHeader("Header"),
        dbcPopoverBody(sprintf("This is a %s popover", placement))
      ),
      id = sprintf("popover-%s", placement),
      target = sprintf("popover-%s-target", placement),
      placement = placement,
      is_open = FALSE
    )
  )
}


make_button <- function(placement) {
  return(
    dbcButton(
      sprintf("Popover on %s", placement),
      id = sprintf("popover-%s-target", placement),
      className = "mx-2",
      n_clicks = 0
    )
  )
}


popovers <- div(
  list(
    make_button("top"), make_button("left"),
    make_button("right"), make_button("bottom"),
    make_popover("top"), make_popover("left"),
    make_popover("right"), make_popover("bottom")
  )
)


toggle_popover <- function(n, is_open) {
  if (n > 0) {
    return(!is_open)
  }
  return(is_open)
}

position <- list("top", "left", "right", "bottom")
for (p in position) {
  app$callback(
    output(sprintf("popover-%s", p), "is_open"),
    list(
      input(sprintf("popover-%s-target", p), "n_clicks"),
      state(sprintf("popover-%s", p), "is_open")
    ),
    toggle_popover
  )
}
